### PR TITLE
Rework parser

### PR DESCRIPTION
**Pull request by SanderJSA**
Current parser backtraces causing evaluation to run in polynomial time, maybe even exponential time in some cases.
We can rework the grammar to run in linear time, thus drastically increasing performance.

### CODE REVIEW

**Comment by CohenArthur** [comment]
Which PR should be merged first? This one or #235 ?

**Comment by SanderJSA** [comment]
This one is still WIP

File: src/instruction/jk_inst.rs [hunk 1]
```diff
@@ -81,27 +81,30 @@ impl Instruction for JkInst {
...
 mod tests {
     use super::*;
-    use crate::parser::Construct;
+    use crate::parser::{Construct, Token};
 
     #[test]
     fn t_invalid_jkinst() {
-        let (_, fc) = Construct::function_call("tamer()").unwrap();
+        let (input, id) = Token::identifier("tamer()").unwrap();
+        let (_, fc) = Construct::function_call(input, &id).unwrap();
```
Comment (Skallwar) [inline src/instruction/jk_inst.rs:89]: The function identifier should be ``tamer``not ``tamer()``,

Comment (SanderJSA) [reply src/instruction/jk_inst.rs:89]: Token::identifier will parse `tamer` leaving the `()` in input which is then fed to function_call.
This is equivalent to giving `tamer()` directly to the previous implementation of function_call.

The reason why identifier has been extracted from function_call is to reduce backtracking.

Comment (SanderJSA) [reply src/instruction/jk_inst.rs:89]: Token::identifier will parse `tamer` leaving the `()` in input which is then fed to function_call.
This is equivalent to giving `tamer()` directly to the previous implementation of function_call.

The reason why identifier has been extracted from function_call is to reduce backtracking.

File: src/parser/constructs.rs [hunk 2]
```diff
@@ -61,8 +61,8 @@ impl Construct {
 
     /// Parse a method call or a field access
     ///
-    /// method_call_or_field_access = instance '.' method_call
-    ///                             | instance '.' field_access
+    /// method_call_or_field_access = instance '.' identifer method_call
+    ///                             | instance '.' identifer field_access
```
Comment (CohenArthur) [inline src/parser/constructs.rs:65]:
```suggestion
    /// method_call_or_field_access = instance '.' identifier method_call
    ///                             | instance '.' identifier field_access
```

Comment (SanderJSA) [reply src/parser/constructs.rs]: Nice catch !

**Review by CohenArthur — APPROVED** [review]
Thanks a lot! Insane amount of work and great PR!

**Pull request merged**